### DISCUSSION TITLE
Converting Hugging Face Safetensor to Checkpoint

### DISCUSSION THREAD

**Discussion by Bojun-Feng** [lean-dojo/ReProver#53] [Q&A]
Hello, I've been trying to reproduce the results of the paper through evaluating the hugging face model on the benchmark (without indexing).

However, I am having some trouble applying the hugging face safetensors to compatible torch lightning check points for evaluation. Currently I am using torch to change the safe tensor to a binary checkpoint:

```
# ... download model from https://huggingface.co/kaiyuy/leandojo-lean4-tacgen-byt5-small
ckpt = load_file('model.safetensors')
torch.save(ckpt, '/content/model.bin')
```

Then, I run command according to the instructions:
```
python prover/evaluate.py 
  --data-path data/leandojo_benchmark_4/random/
  --ckpt_path /content/model.bin
  --split test
  --num-workers 5
  --num-gpus 1
```

However, it seems that dictionary keys are missing in the resulting checkpoint, I get a dictionary error for 'pytorch_lightning_version'.

I first manually added the pytorch lightning version, but am not sure what the parameter should be for other missing parameters such as 'max_inp_seq_len', and 'max_oup_seq_len'. Is there a conveinent way I can load the model from the safetensors or gain access to the original model checkpoints? I looked online but found no relevant information.

I am new to the repo and lean in general so this might be a stupid question. Would appreaciate any useful information or suggestions.

**Comment by yangky11** [comment]
You probably want to use the PyTorch Lightning checkpoints here: https://huggingface.co/kaiyuy/leandojo-pl-ckpts

**Reply by xiaoxin-yin** [reply]
Thanks, Kaiyu. 

I am trying to run evaluate.py with ReProver. I first ran "python retrieval/index.py" to generate indexed corpus, which seems to be successful. Then I use the checkpoints from https://huggingface.co/kaiyuy/leandojo-pl-ckpts to run ReProver, as follows:

python prover/evaluate.py \
--data-path ./data/leandojo_benchmark_4/novel_premises/  \
--ckpt_path ./ckpt/leandojo-pl-ckpts/reprover_random.ckpt \
--indexed-corpus-path ./data/indexed_corpus \
--split test --num-workers 1 --num-gpus 1

However, it produces the following error:


[2024-05-13 22:13:24,303] [INFO] [real_accelerator.py:203:get_accelerator] Setting ds_accelerator to cuda (auto detect)
 [WARNING]  async_io requires the dev libaio .so object and headers but these were not found.
 [WARNING]  async_io: please install the libaio-dev package with apt
 [WARNING]  If libaio is already installed (perhaps from source), try setting the CFLAGS and LDFLAGS environment variablesto where it can be found.
 [WARNING]  Please specify the CUTLASS repo directory as environment variable $CUTLASS_PATH
 [WARNING]  sparse_attn requires a torch version >= 1.5 and < 2.0 but detected 2.1
 [WARNING]  using untested triton version (2.1.0), only 1.0.0 is known to be compatible
/home/mcwave/anaconda3/envs/atp/lib/python3.10/site-packages/trio/_core/_multierror.py:468: RuntimeWarning: You seem to already have a custom sys.excepthook handler installed. I'll skip installing Trio's custom handler, but this means MultiErrors will not show full tracebacks.
  warnings.warn(
2024-05-13 22:13:24.792 | INFO     | __main__:main:218 - PID: 522
2024-05-13 22:13:24.792 | INFO     | __main__:main:219 - Namespace(data_path='/home/mcwave/code/3rdparty/ReProver/data/leandojo_benchmark_4/novel_premises/', exp_id=None, split='test', file_path=None, full_name=None, name_filter=None, num_theorems=None, ckpt_path='/home/mcwave/code/3rdparty/ReProver/ckpt/leandojo-pl-ckpts/reprover_novel_premises.ckpt', indexed_corpus_path='/home/mcwave/code/3rdparty/ReProver/data/indexed_corpus', tactic=None, module=None, num_sampled_tactics=64, timeout=600, num_workers=1, num_gpus=1, verbose=False)
2024-05-13 22:13:38.627 | INFO     | __main__:_get_theorems_from_files:83 - 2000 theorems loaded from /home/mcwave/code/3rdparty/ReProver/data/leandojo_benchmark_4/novel_premises/
model_cls <class 'generator.model.RetrievalAugmentedGenerator'>
Deepspeed: /home/mcwave/code/3rdparty/ReProver/ckpt/leandojo-pl-ckpts/reprover_novel_premises.ckpt
convert_zero_checkpoint_to_fp32_state_dict: /home/mcwave/code/3rdparty/ReProver/ckpt/leandojo-pl-ckpts/reprover_novel_premises.ckpt to /tmp/tmp14ael687/lightning.cpkt
Processing zero checkpoint '/home/mcwave/code/3rdparty/ReProver/ckpt/leandojo-pl-ckpts/reprover_novel_premises.ckpt/checkpoint'
Detected checkpoint of type zero stage 2, world_size: 1
Parsing checkpoint created by deepspeed==0.14.0
Reconstructed Frozen fp32 state dict with 111 params 217657472 elements
Reconstructed fp32 state dict with 170 params 299637760 elements
Saving fp32 state dict to /tmp/tmp14ael687/lightning.cpkt
load_from_checkpoint: /tmp/tmp14ael687/lightning.cpkt
2024-05-13 22:13:43.991 | INFO     | generator.model:__init__:113 - Loading the retriever from lightning_logs/retriever_novel_premises/checkpoints/epoch=3-step=172908.ckpt/
model_cls <class 'retrieval.model.PremiseRetriever'>

---------------------------------------------------------------------------------------------------------------------------
evaluate.py 244 <module>
main()

evaluate.py 221 main
pass_1 = evaluate(

evaluate.py 116 evaluate
prover = DistributedProver(

proof_search.py 397 __init__
tac_gen = RetrievalAugmentedGenerator.load(

model.py 131 load
return load_checkpoint(cls, ckpt_path, device, freeze)

common.py 476 load_checkpoint
model = model_cls.load_from_checkpoint(path, strict=False)

model_helpers.py 125 wrapper
return self.method(cls, *args, **kwargs)

module.py 1581 load_from_checkpoint
loaded = _load_from_checkpoint(

saving.py 91 _load_from_checkpoint
model = _load_state(cls, checkpoint, strict=strict, **kwargs)

saving.py 158 _load_state
obj = cls(**_cls_kwargs)

model.py 114 __init__
self.retriever = PremiseRetriever.load(

model.py 51 load
return load_checkpoint(cls, ckpt_path, device, freeze)

common.py 466 load_checkpoint
if not _is_deepspeed_checkpoint(ckpt_path):

common.py 459 _is_deepspeed_checkpoint
raise FileExistsError(f"Checkpoint {path} does not exist.")

FileExistsError:
Checkpoint lightning_logs/retriever_novel_premises/checkpoints/epoch=3-step=172908.ckpt/ does not exist.


It seems to convert the original checkpoint into a lightning checkpoint in the /tmp folder. But then it tries to access lightning_logs/retriever_novel_premises/checkpoints/epoch=3-step=172908.ckpt

**Reply by Bojun-Feng** [reply]
Thanks for the response @yangky11! I tried out your suggestion and am facing the same issues as @xiaoxin-yin.

Additionally, there are quite a few different models in the checkpoint, and I am not sure what they represent. What model should I use for benching without indexed data? Also, the checkpoints does not seem to align with the hugging face models and seem to be correlated with different datasets. Is it because the checkpoints already contain the premises?

I'm new to leandojo and lean so I'm sorry if my questions are silly. After I figure out how to benchmark with the checkpoints, I would be happy to write up some documentation and a demo colab notebook for future users!

**Reply by yangky11** [reply]
It looks like the error is because the ReProver's checkpoint contains a path to the retriever checkpoint, which points to the path when I trained the retriever (`lightning_logs/retriever_novel_premises/checkpoints/epoch=3-step=172908.ckpt`). Could you please try renaming the `reprover_novel_premises.ckpt` folder in https://huggingface.co/kaiyuy/leandojo-pl-ckpts to `lightning_logs/retriever_novel_premises/checkpoints/epoch=3-step=172908.ckpt`?

**Reply by yangky11** [reply]
@Bojun-Feng 

* [generator_novel_premises.ckpt](https://huggingface.co/kaiyuy/leandojo-pl-ckpts/tree/main/generator_novel_premises.ckpt): Tactic generator (without retrieval) trained on the `novel_premises` split of LeanDojo Benchmark.
* [generator_random.ckpt](https://huggingface.co/kaiyuy/leandojo-pl-ckpts/tree/main/generator_random.ckpt): Tactic generator (without retrieval) trained on the `random` split of LeanDojo Benchmark.
* [retriever_random.ckpt](https://huggingface.co/kaiyuy/leandojo-pl-ckpts/tree/main/retriever_random.ckpt): Retriever trained on the `random` split.
* [retriver_novel_premises.ckpt](https://huggingface.co/kaiyuy/leandojo-pl-ckpts/tree/main/retriver_novel_premises.ckpt): Retriever trained on the `novel_premises` split.
* [reprover_novel_premises.ckpt](https://huggingface.co/kaiyuy/leandojo-pl-ckpts/tree/main/reprover_novel_premises.ckpt): Retrieval-augmented tactic generator trained on `novel_premises`.
* [reprover_random.ckpt](https://huggingface.co/kaiyuy/leandojo-pl-ckpts/tree/main/reprover_random.ckpt): Retrieval-augmented tactic generator trained on `random`.

**Reply by xiaoxin-yin** [reply]
Thank you, @yangky11 ! It works after I rename the checkpoint to lightning_logs/retriever_novel_premises/checkpoints/epoch=3-step=172908.ckpt. However, I found that checkpoints like "kaiyuy/leandojo-lean4-retriever-tacgen-byt5-small" work better than those in https://huggingface.co/kaiyuy/leandojo-pl-ckpts. Just wondering if it is expected.

**Reply by yangky11** [reply]
I expect them to be the same models, only in different formats.

**Comment by yangky11** [comment]
Update: The new evaluation script takes Hugging Face (instead of PyTorch Ligthning) checkpoints as input. You can directly use models such as [leandojo-lean4-tacgen-byt5-small](https://huggingface.co/kaiyuy/leandojo-lean4-tacgen-byt5-small). Please see the updated README for details. As a result, https://huggingface.co/kaiyuy/leandojo-pl-ckpts has been deleted.
